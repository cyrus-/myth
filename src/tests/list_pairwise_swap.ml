let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

let list_pairwise_swap : list -> list |>
{
  [] => []
| [0] => []
| [1] => []
| [2] => []
| [2;2] => [2;2]
| [0;1] => [1;0]
| [1;0] => [0;1]
| [1;2] => [2;1]
| [2;1] => [1;2]
| [0;2] => [2;0]
| [0;1;0] => []
| [1;0;1] => []
| [2;0;2] => []
| [1;2;1] => []
| [2;1;2] => []
| [0;1;0;1] => [1;0;1;0]
| [1;0;1;0] => [0;1;0;1]
| [1;2;1;2] => [2;1;2;1]
| [2;1;2;1] => [1;2;1;2]
| [0;2;0;2] => [2;0;2;0]
} = ?"
