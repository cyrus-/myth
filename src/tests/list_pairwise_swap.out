(* Generated Thu Nov 13 00:35:51 2014 *)
let list_pairwise_swap : list -> list =
  let rec f1 (l1:list) : list =
    match l1 with
      | Nil -> Nil
      | Cons (n1, l2) -> (match f1 l2 with
                            | Nil -> (match l2 with
                                        | Nil -> Nil
                                        | Cons (n2, l3) -> Cons (n2,
                                                             Cons (n1, f1 l3)))
                            | Cons (n2, l3) -> Nil)
  in
    f1
;;
